src/main/scala/is/hail/expr/AST.scala

```diff
@@ -892,7 +893,7 @@ case class Let(posn: Position, bindings: Array[(String, AST)], body: AST) extend
   def toIR(agg: Option[String] = None): Option[IR] = for {
     irBindings <- anyFailAllFail(bindings.map { case (x, y) => y.toIR(agg).map(irY => (x, irY)) })
     irBody <- body.toIR(agg)
-  } yield irBindings.foldRight(irBody) { case ((name, v), x) => ir.Let(name, v, x, x.typ) }
```

Comment:
Curious, why did you take the type out of the let?

Reply:
sometimes x.typ isn't well defined, e.g. for an ArrayMap of something where the type of `body` hasn't been inferred yet. Calling `.typ` on that causes an error in TArray, because you can't have TArray(null). I guess I could have tried to take this the other way and made sure that all IRs generated with toIR had a well-defined type. I think I thought that was going to be more difficult/dicier, but I can't recall why.

Reply:
Yeah, you're right, certainly simpler to let Infer do its job.  If you reorg the primOp function a bit, I think this will be good to go.

Reply:
ok, done

Reply:
Thanks!